include/date/tz.h

```diff
@@ -2183,42 +2197,224 @@ from_stream(std::basic_istream<CharT, Traits>& is, const CharT* fmt,
     return is;
 }
 
-template <class Duration>
-inline
-sys_time<typename std::common_type<Duration, std::chrono::seconds>::type>
-to_sys_time(const tai_time<Duration>& t)
+template<typename SourceClock, typename DestClock>
+struct clock_time_conversion
+{};
+
+template<>
+struct clock_time_conversion<std::chrono::system_clock, std::chrono::system_clock>
```

Comment:
I think we can get rid of the identity conversions for `system_clock` and `utc_clock` as they can be handled by the generic identity conversion.  This simplifies us down to 3 `clock_time_conversion` specializations.

Reply:
This would make ```clock_time_conversion<utc_clock, utc_clock>``` ambigous, as following candidates would be equal matches:
```
clock_time_conversion<Clock, Clock>
clock_time_conversion<Clock, utc_clock>
clock_time_conversion<utc_clock, Clock>
```
The same goes for ``std::chrono::system_clock``.

Reply:
At least on gcc-6.2 they are ambiguous (and I tend to agree with this interpretation).

Reply:
Interesting.  Looks like we found a difference in behavior between gcc and clang.  We'll keep 'em and let the compilers work at that difference.

Reply:
Hmmm.. Shouldn't `clock_time_conversion<Clock, utc_clock>` be SFINAE'd out if `Clock::to_utc(utc_time<Duration>)` doesn't exist?

Reply:
The design behing the `clock_time_conversion<S, D>` is that the specialization is always valid, and only `operator()` is SFINAE'd. This apply to generic specialization (that does not have it at all), but also to `clock_time_conversion<C, C>` that will be valid for `C` being `int`, but the `operator()` will accept only `time_point<C>` - potentially this could be an identity.

Before, this specialization (`<S, utc_clock>`, `<utc_clock, D>`, `<S, sys_cloc>`, `<sys_clock, D>`) were SFINAE'd, but this workaround for compiler making code ill-formed if `S` was not having specific function (like `from_sys`). I have added the `return_to_sys` (and other) traits to avoid needs to SFINAE specialization and make them consistent.

Finally, there was not easy answer, what `Duration` should be used to SFINAE whole trait.